### ISSUE TITLE
Provide a mechanism for registering a KVO observer

### ISSUE THREAD

**Issue by stuartmorgan** [dart-lang/native#1508]
`NSObject`'s `addObserver:forKeyPath:options:context:` allows registering an observer for arbitrary changes to an object, where the observer can be any object implementing the `observeValueForKeyPath:ofObject:change:context:` method. This is a strange API in that the normal way of doing this would be for the `addObserver:` parameter to be a protocol, but instead it's all just informal duck-typing: it takes `NSObject`, and it's up to the caller to pass an object that meets the requirements, with no compiler help. Because there's no protocol, the generator support for protocols doesn't kick in, and AFAICT there's no way to use this method via FFI without manually writing a block adapter class.

It would be nice if there were some built-in support for this since KVO support is a pretty common feature of Apple SDK classes, so that clients wouldn't have to dip into writing native code to access it. Maybe ffigen could hard-code a pseudo-protocol that it would generate for this? Or `objective_c` could have something hand-written for it?

**Comment by liamappelbe** [comment]
First question is what the observers look like. Classes implementing some `Observer` interface, or just `Function`s? `Function`s would be more convenient IMO. See API suggestions below.

### Memory management

Memory management is going to be the main complication here. From my reading, it looks like the observed object holds a weak reference to the observer. But it's not an auto-zeroing weak reference like you get in ARC, so you get a use-after-free error (typically a seg fault) if the observer is deleted then the observed tries to send it an update.

The KVO contract is that it's up to the user to deregister the observer before it is destroyed (often during the destructor). But Dart doesn't give users an easy way of doing this sort of finalization work, so we'll have to solve this in a different way. Couple of possibilities:

1)  **Weak refs**: The closest solution to ObjC's would be for the observed to continue holding weak refs to the observers, and do some native finalization magic to automatically deregister the observer when it's GC'd. Of course, that would mean the user is now responsible for making sure the observer stays in scope as long as they need it. And the failure mode if they forget to do that is they simply don't see the updates, which would be very confusing.
2) **Strong refs**: If the observed holds a strong ref to the observer, there's no possibility of use-after-free, no need for finalization magic, and no need for the user to manually hold strong references to the observer. But it would be very easy to accidentally create a reference cycle this way, especially if the observer is a function. Implementation details TBD, but such a cycle would almost certainly prevent the observed object from being GC'd. It's probably a good idea to pass the observed object as a parameter to the observer, to reduce the temptation to create a reference cycle. That plus good documentation might be the best option.

I prefer option 2. In either case we'd still provide a manual deregistration method.

### API ideas

```Dart
typedef Observer<O extends NSObject, T extends ObjCObjectBase> =
    void Function(O target, T oldValue, T newValue);

extension Observed on NSObject {
  Observation addObserver<O, T>(Observer<O, T> observer, String key);
}

class Observation {
  void remove();
}
```

Since we're doing codegen anyway, we could make this a bit more typesafe by generating the `addObserver` method on every subclass of `NSObject`:

```Dart
class Foo extends NSObject {
  Observation addObserver<Foo, T>(Observer<Foo, T> observer, String key);
}
```

Or if we want to be reeeeealy type safe, we could code gen an `addObserver` method per observable property. But idk how practical that is. If there are observable keys that don't correspond 1:1 with publicly visible properties in the API, then this won't work.

```Dart
class Foo extends NSObject {
  Bar get baz;

  Observation addObserver_baz<Foo, Bar>(Observer<Foo, Bar> observer);
}
```

I lean towards the middle option, with one `addObserver` method per class.

**Comment by stuartmorgan-g** [comment]
> `Function`s would be more convenient IMO.

They are more convenient, but they aren't the Obj-C interface. The Obj-C equivalent of that would be a KVO system using blocks, which isn't something Apple ever added. Isn't the goal at the layer of this library to expose the underlying APIs as directly as possible, not to re-imagine them in a more Dart-like way?

> Of course, that would mean the user is now responsible for making sure the observer stays in scope as long as they need it.

Which is exactly how KVO works in Obj-C.

> And the failure mode if they forget to do that is they simply don't see the updates, which would be very confusing.

Isn't this only confusing if you use functions instead of mirroring the Obj-C API?

> If the observed holds a strong ref to the observer, there's no possibility of use-after-free, no need for finalization magic, and no need for the user to manually hold strong references to the observer. But it would be very easy to accidentally create a reference cycle this way

The memory semantics mismatch between Obj-C and Dart and the potential for cycles is already a significant foot-gun with ffigen IMO; taking an Obj-C API and mirroring it with the *opposite* of the underlying APIs memory semantics seems incredibly dangerous.

**Comment by liamappelbe** [comment]
> Isn't the goal at the layer of this library to expose the underlying APIs as directly as possible, not to re-imagine them in a more Dart-like way?

Not necessarily. I'm trying to make things a bit more ergonomic than a pure translation of the ObjC API, while still being familiar to ObjC developers (eg, adding methods to the ObjC collections to match the Dart collections). In this case, at the very least we're going to need to change the memory management semantics of the ObjC API, to automatically call `removeObserver` when Dart GCs the observer, since Dart doesn't have an easy way for users to write destructors/finalizers (also there would be threading issues if the dealloc method was written in Dart).

> Isn't this only confusing if you use functions instead of mirroring the Obj-C API?

I don't think so. I think it would be just as confusing if it were an interface. In this scenario, the user has created their observer implementation, called `addObserver`, then dropped references to the observer. They may receive some notifications, but at some point the object will be GC'd and the observations will just stop. It's still a confusing failure mode that would be hard for users to debug.

### Interface approach

Fleshing out your suggestion a bit more, where we try to more directly mirror the ObjC approach, using interfaces and keeping the weak reference semantics. I'd want to force the user to implement the `observeValueForKeyPath:ofObject:change:context:` method. So I'd need to expose this as a protocol:

```Dart
// Generated by ffigen from a protocol defined in ObjC.
// This protocol probably already exists, otherwise I'll just write it myself.
interface class Observer extends ObjCProtocolBase implements NSObjectProtocol {
  static Observer implement(
      {required void Function(NSString, ObjCObjectBase, NSDictionary, Pointer<Void>) observeValueForKeyPath,
      bool $keepIsolateAlive = true}) {
    ...
  }
  ...
}
```

Then I can add an extension on `NSObject`. This looks almost identical to the existing `addObserver` method, except that the `observer` arg has the correct type. I'll rename/hide that existing method, then this method can delegate to it.

```Dart
extension Observed on NSObject {
  void addObserver(Observer observer,
      {required NSString forKeyPath,
      required NSKeyValueObservingOptions options,
      required Pointer<Void> context) {
    ...
  }

  void removeObserver(Observer observer,
      {required NSString forKeyPath,
      required Pointer<Void> context) {
    ...
  }
}
```

That's all pretty straightforward. The real issue is how this API interacts with memory management. We need to auto-remove the observer when it's GC'd, otherwise there will be seg faults if the user doesn't manage the lifecycle correctly (though I suppose it's arguable that an obvious seg fault is better than silently not receiving change notifications). Since an observer may be passed to multiple `addObserver` calls, it needs to maintain a list/set of the observations (`(object, keyPath, context)` triples) it's observing (the `object` ref in the triple should be a strong ref).

I can store the list of observations on the protocol implementation using `objc_setAssociatedObject`. I need to use associated objects here, rather than writing a wrapper object around the `Observer` protocol, because the user is responsible for holding the observer for as long as necessary, and this would mean they need to hold a reference to the wrapper object rather than the `Observer` protocol they created. I also considered creating an `Observation` wrapper object around the `Observer` protocol in each `addObserver` call, which would be much simpler, but runs into the same lifecycle issues.

So users would just have to implement the protocol using `Observer.implement`/`implementAsListener` etc, call `addObserver`, and then hold a reference to the `Observer` for as long as they need to receive change notifications.

### More options

So now we have two extremes, one that looks very much like pure ObjC, and one that's a lot more Darty. But there are other options in between. I think there are 3 mostly orthogonal design questions:

1) Interface vs function?
2) Weak vs strong refs? Does the user need to hold a reference to something to keep the observation active? Or is the observer kept alive by the observed object (or until `removeObserver` is called)?
3) Observation token? Does the `addObserver` function return an object representing the observation?

The Darty approach is function + strong + token, whereas the ObjC approach is interface + weak + no token.

Another data point is [Swift's approach](https://developer.apple.com/documentation/swift/using-key-value-observing-in-swift), where the `observe` method still takes an interface, but returns a `NSKeyValueObservation` which must be held for as long as the user needs to observe changes (they don't have to directly hold a reference to the observer afaict). So that's interface + weak + token. This would mean the API looks slightly different from the pure ObjC API, but I think it's a nice middle ground.

**Comment by stuartmorgan-g** [comment]
> > Isn't this only confusing if you use functions instead of mirroring the Obj-C API?
> 
> I don't think so. I think it would be just as confusing if it were an interface. In this scenario, the user has created their observer implementation, called `addObserver`, then dropped references to the observer. They may receive some notifications, but at some point the object will be GC'd and the observations will just stop. It's still a confusing failure mode that would be hard for users to debug.

I don't understand; why would someone drop references to the observer when they still want to observe? I don't understand how this is any different from Obj-C, where it is your responsibility to keep the observing object alive for as long as you want to observe things.

It sounds like you are expecting the observer to be a throw-away object; I have never seen KVO used that way in all my time doing Obj-C development. The observer is some important object that needs information. and so the need to observe and the lifetime of the object observing are inherently tied together, such that the weak refs don't actually cause any problems.

> The Darty approach is function + strong + token, whereas the ObjC approach is interface + weak + no token.

I really don't think making a random-to-clients subset of the generated interface a strongly opinionated, non-Obj-C-style API while the rest is a more or less direct mirror of Obj-C, is going to lead to good outcomes. If most of the API surface requires a reasonable understanding of Obj-C to use correctly, then the target of this layer should be people familiar with Obj-C development. Having pockets that completely defy their expectations will mean that *nobody* will have a good understanding of how to use the generated APIs.

For instance, I would absolutely expect that if everything else stays basically as is, but *just* KVO reverses the memory semantics, then basically everyone who tries to use KVO will leak everything by accident.

> Another data point is [Swift's approach](https://developer.apple.com/documentation/swift/using-key-value-observing-in-swift), where the `observe` method still takes an interface, but returns a `NSKeyValueObservation` which must be held for as long as the user needs to observe changes (they don't have to directly hold a reference to the observer afaict). So that's interface + weak + token. This would mean the API looks slightly different from the pure ObjC API, but I think it's a nice middle ground.

I'm not familiar with Swift KVO, but following a Swift model would be much better than inventing a new Dart-y model, IMO, for the reasons above.

**Comment by liamappelbe** [comment]
Just to clarify, I'm not arguing for [this approach](https://github.com/dart-lang/native/issues/1508#issuecomment-2915030014). Your comment [here](https://github.com/dart-lang/native/issues/1508#issuecomment-2934680022) convinced me that I needed to explore the options more carefully. So that's what I'm doing now. I appreciate your input.

> I don't understand; why would someone drop references to the observer when they still want to observe? I don't understand how this is any different from Obj-C, where it is your responsibility to keep the observing object alive for as long as you want to observe things.
> 
> It sounds like you are expecting the observer to be a throw-away object; I have never seen KVO used that way in all my time doing Obj-C development. The observer is some important object that needs information. and so the need to observe and the lifetime of the object observing are inherently tied together, such that the weak refs don't actually cause any problems.

Since Dart is a GC'd language, I have a hunch that users aren't as used to reasoning about object lifecycles as they are in more manual languages like ObjC/C++/Rust etc. Personally I have a very different frame of mind when I'm coding in a GC'd language vs a non GC language.

So I'm not necessarily expecting the observer itself to be short lived. Under a GC mindset I'm expecting that if something is using the observer, it will be kept alive. This is how it works with callbacks (whether they're functions or interfaces) throughout the rest of Dart's APIs.

---

Focusing on the weak ref approach, what do you think should happen if the user doesn't manage the observer's lifecycle correctly, and it's GC'd before the observed object without calling `removeObserver`? In ObjC, subsequent update notifications would cause seg faults (but it's easier to solve the problem since you can just write a `dealloc` method). Should I automatically call `removeObserver` when the observer is GC'd?

**Comment by stuartmorgan-g** [comment]
> Should I automatically call `removeObserver` when the observer is GC'd?

Yes, provided there's a reasonable way to track whether it's already been done. It is (unfortunately) an error to remove an observer that's not set.

> Since Dart is a GC'd language, I have a hunch that users aren't as used to reasoning about object lifecycles as they are in more manual languages like ObjC/C++/Rust etc. Personally I have a very different frame of mind when I'm coding in a GC'd language vs a non GC language.

The problem is—and this is a general concern I have with Obj-C FFI—clients of these APIs fundamentally *aren't* coding in a GC'd language. They are coding in a *hybrid* of two systems, one of which is GC'd and one of which is not, and there are complex interactions that people will have to understand. People who think they are just writing regular Dart code are going to have a bad time. People who know Dart, don't know Obj-C, and use this system without being able to reason about object lifecycles are going to have a bad time. Unless we are going to completely solve that problem (and I don't see how we can), I don't think we should try to obfuscate it in a subset of the surfaces.

**Comment by liamappelbe** [comment]
Made a [PR](https://github.com/dart-lang/native/pull/2344) with a prototype of swift style observers. You can take a look at [observer_test.dart](https://github.com/dart-lang/native/pull/2344/files#diff-4a9a6f19024b2ed4a6fba1504e4bf626a83c82b74604c969ce38259104d283fa) to see how it looks from a user's perspective.
